catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/MigrationTaskManager.java

```diff
@@ -0,0 +1,194 @@
+     * of {@link MigrationFileWriter} for processing.
+     *
+     * @param results          The results of a catalog query that need to be exported.
+     * @param exportGroupCount The group or page number of the query. This value is not monitored
+     *                         or audited for repeats and is strictly for record keeping by naming
+     *                         the resulting export files appropriately.
+     * @throws MigrationException Thrown if one of the writing threads fails or throws an exception
+     *                            itself.
+     */
+    public void exportMetacardQuery(final List<Result> results, long exportGroupCount)
+            throws MigrationException {
+        for (int i = 0; i < results.size(); i += catalogConfig.getExportCardsPerFile()) {
+            final List<Result> fileResults = results.subList(i,
+                    Math.min((i + catalogConfig.getExportCardsPerFile()), results.size()));
+            final File exportFile = catalogConfig.getExportPath()
+                    .resolve(makeFileName(exportGroupCount, i))
+                    .toFile();
+
+            CatalogWriterCallable writerCallable = createWriterCallable(exportFile,
+                    fileResults,
+                    fileWriter);
+
+            ListenableFuture<Void> task = taskExecutor.submit(writerCallable);
+
+            Futures.addCallback(task, createFutureCallback());
+        }
+
+        checkForFailures();
+
+        //        List<Future<Void>> finishedTasks = new ArrayList<>();
+        //
+        //        for (Future<Void> possibleFinishedTask : taskData) {
+        //            if (possibleFinishedTask.isDone()) {
+        //                finishedTasks.add(possibleFinishedTask);
+        //            }
+        //        }
+        //
+        //        for (Future<Void> finishedTask : finishedTasks) {
+        //            try {
+        //                finishedTask.get();
+        //            } catch (ExecutionException | InterruptedException | RuntimeException e) {
+        //                LOGGER.error("Task in manager encountered an error: {}", e.getMessage(), e);
+        //                throw new MigrationException("Catalog could not export metacards", e);
+        //            } finally {
+        //                taskData.remove(finishedTask);
+        //            }
+        //        }
+    }
+
+    AtomicReference<Throwable> getAtomicReference() {
+        return taskFailure;
+    }
+
+    FutureCallback<Void> createFutureCallback() {
+        return new FutureCallback<Void>() {
+            @Override
+            public void onSuccess(Void aVoid) {
+                LOGGER.debug("File write complete");
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                LOGGER.error("File writing thread threw an exception");
+                taskFailure.set(throwable);
```

Comment:
If we let the callback handle success/error, our exception we throw would be ignored since the callback is only executed on the executor service. Also - there's no point to continue processing when there's an error, so we check the atomic reference periodically and if an executor service thread threw an error, our main thread throws a migration exception.